src/Compilers/Core/Portable/MetadataReader/MetadataHelpers.cs

```diff
@@ -556,22 +581,30 @@ internal static string UnmangleMetadataNameForArity(string emittedTypeName, int
         /// An ImmutableArray representing the single string "System"
         /// </summary>
         private static readonly ImmutableArray<string> s_splitQualifiedNameSystem = ImmutableArray.Create(SystemString);
+        private static readonly ImmutableArray<ReadOnlyMemory<char>> s_splitQualifiedNameSystemMemory = ImmutableArray.Create(SystemString.AsMemory());
 
-        internal static ImmutableArray<string> SplitQualifiedName(
-              string name)
+        internal static ImmutableArray<string> SplitQualifiedName(string name)
+            => SplitQualifiedNameWorker(name.AsMemory(), s_splitQualifiedNameSystem, static memory => memory.ToString());
```

Comment:
>memory.ToString() [](http://example.com/codeflow?start=102&length=17)

Probably should reuse original string when possible.

Reply:
so in this case, we want to reuse the same immutablearray and not allocate a new one (in either the memory or string case).  so that's why each pass in the singleton to use in that case.